src/amalgam/gen/neondot.c

```diff
@@ -5958,3 +5962,160 @@ void xnn_qu8_igemm_minmax_rndnu_ukernel_4x8c4__neondot(
+              const int8x16_t rhs_vec_2_0 = vshlq_n_s8(rhs_raw_vec_2, 4);
+              const int8x16_t rhs_vec_3_0 = vshlq_n_s8(rhs_raw_vec_3, 4);
+              const int8x16_t rhs_vec_4_0 = vshlq_n_s8(rhs_raw_vec_4, 4);
+              const int8x16_t rhs_vec_5_0 = vshlq_n_s8(rhs_raw_vec_5, 4);
+              const int8x16_t rhs_vec_6_0 = vshlq_n_s8(rhs_raw_vec_6, 4);
+              const int8x16_t rhs_vec_7_0 = vshlq_n_s8(rhs_raw_vec_7, 4);
+
+              // High nibble
+              const int8x16_t rhs_vec_0_1 = vandq_s8(rhs_raw_vec_0, nibble_mask);
+              const int8x16_t rhs_vec_1_1 = vandq_s8(rhs_raw_vec_1, nibble_mask);
+              const int8x16_t rhs_vec_2_1 = vandq_s8(rhs_raw_vec_2, nibble_mask);
+              const int8x16_t rhs_vec_3_1 = vandq_s8(rhs_raw_vec_3, nibble_mask);
+              const int8x16_t rhs_vec_4_1 = vandq_s8(rhs_raw_vec_4, nibble_mask);
+              const int8x16_t rhs_vec_5_1 = vandq_s8(rhs_raw_vec_5, nibble_mask);
+              const int8x16_t rhs_vec_6_1 = vandq_s8(rhs_raw_vec_6, nibble_mask);
+              const int8x16_t rhs_vec_7_1 = vandq_s8(rhs_raw_vec_7, nibble_mask);
+
+              const int8x16_t lhs_vec_0 = vld1q_s8((const int8_t*)(lhs_ptr + 0));
+              const int8x16_t lhs_vec_1 = vld1q_s8((const int8_t*)(lhs_ptr + 16));
+
+              lhs_ptr += 32;
+              rhs_ptr += 128;
+
+              int8x16_t t;
+
+              t = vcombine_s8(vget_low_s8(lhs_vec_0), vget_low_s8(lhs_vec_0));
+              iacc0011 = vdotq_s32(iacc0011, rhs_vec_0_0, t);
+              iacc2233 = vdotq_s32(iacc2233, rhs_vec_1_0, t);
+              iacc4455 = vdotq_s32(iacc4455, rhs_vec_2_0, t);
+              iacc6677 = vdotq_s32(iacc6677, rhs_vec_3_0, t);
+              t = vcombine_s8(vget_high_s8(lhs_vec_0), vget_high_s8(lhs_vec_0));
+              iacc0011 = vdotq_s32(iacc0011, rhs_vec_4_0, t);
+              iacc2233 = vdotq_s32(iacc2233, rhs_vec_5_0, t);
+              iacc4455 = vdotq_s32(iacc4455, rhs_vec_6_0, t);
+              iacc6677 = vdotq_s32(iacc6677, rhs_vec_7_0, t);
+
+              t = vcombine_s8(vget_low_s8(lhs_vec_1), vget_low_s8(lhs_vec_1));
+              iacc0011 = vdotq_s32(iacc0011, rhs_vec_0_1, t);
+              iacc2233 = vdotq_s32(iacc2233, rhs_vec_1_1, t);
+              iacc4455 = vdotq_s32(iacc4455, rhs_vec_2_1, t);
+              iacc6677 = vdotq_s32(iacc6677, rhs_vec_3_1, t);
+              t = vcombine_s8(vget_high_s8(lhs_vec_1), vget_high_s8(lhs_vec_1));
+              iacc0011 = vdotq_s32(iacc0011, rhs_vec_4_1, t);
+              iacc2233 = vdotq_s32(iacc2233, rhs_vec_5_1, t);
+              iacc4455 = vdotq_s32(iacc4455, rhs_vec_6_1, t);
+              iacc6677 = vdotq_s32(iacc6677, rhs_vec_7_1, t);
+          }
+
+          int32x4_t iacc0 = vpaddq_s32(iacc0011, iacc2233);
+          int32x4_t iacc1 = vpaddq_s32(iacc4455, iacc6677);
+
+          // LHS offset
+          const int32x4_t lhs_offset = vld1q_dup_s32((const int32_t*)lhs_ptr);
```

Comment:
These should be in quantization params?
Note we are thinking about adding a 3rd param - a zero point.  Which is awkwardly 3 floats.
I wish we had 3 arrays instead of an array of offset, scale, zero point, but on ARM we can use LD3